### DISCUSSION TITLE
Usage of the 'collection append' operation warning when use @event.listen

### DISCUSSION THREAD

**Discussion by wangxin688** [sqlalchemy/sqlalchemy#9501] [Usage Questions] [answered]
I want to record the db object change, so I created a Audit log mixin to record the database change to object to reduce a lot of duplicated code. I just check the examples you guys wrote before and made this mixin, but when I use it, there will be a warning generated.
`SAWarning: Usage of the 'collection append' operation is not currently supported within the execution stage of the flush process. Results may not be consistent.  Consider using alternative event listeners or connection-level operations instead.
  target.audit_log.append(`
but the audit log was successfully add to database. should I just ignoring this message? or there is a better to way to do this?
here's my code, I did not paste all of code here. (by the way, this code is also working under async session, amazing :--)
```python3
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Dict, List, Optional

from sqlalchemy import DateTime, ForeignKey, String, event, inspect
from sqlalchemy.dialects.postgresql import ARRAY, HSTORE, JSON
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import Mapped, class_mapper, mapped_column, relationship
from sqlalchemy.orm.attributes import get_history
from sqlalchemy.sql import func

from src.db.db_base import Base
from src.db.db_types import GUID, GUID_GEN
from src.register.context import auth_user_ctx, request_id_ctx, tenant_ctx

if TYPE_CHECKING:
    from src.app.auth.models import User


def get_object_changes(obj):
    """Given a model instance, returns dict of pending
    changes waiting for database flush/commit.

    e.g. {
        'some_field': {
            'before': *SOME-VALUE*,
            'after': *SOME-VALUE*
        },
        ...
    }
    """
    inspection = inspect(obj)
    changes = {}
    for attr in class_mapper(obj.__class__).column_attrs:
        if getattr(inspection.attrs, attr.key).history.has_changes():
            if get_history(obj, attr.key)[2]:
                before = get_history(obj, attr.key)[2].pop()
                after = getattr(obj, attr.key)
                if before != after:
                    if before or after:
                        changes[attr.key] = {"before": before, "after": after}
    return changes


class AuditLog:
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, server_default=GUID_GEN)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=func.now(), nullable=False)
    request_id: Mapped[str] = mapped_column(String(50), nullable=False, comment="request_context_id")
    action: Mapped[str] = mapped_column(String(10), nullable=False, comment="literal action")
    change_data: Mapped[dict] = mapped_column(JSON, nullable=False, comment="object change data")

    @declared_attr
    def user_id(cls):
        return mapped_column(
            GUID, ForeignKey("auth_user.id", ondelete="SET NULL"), default=auth_user_ctx.get, nullable=True
        )

    @declared_attr
    def auth_user(cls) -> Mapped["User"]:
        return relationship("User")


class AuditLogMixin:
    """Use generic association which persists association objects within
    individual tables. each one generates to persist those objects on behalf of
    a particular parent class.
    """

    @declared_attr
    def audit_log(cls) -> Mapped[List["AuditLog"]]:
        cls.AuditLog = type(
            "%sAuditLog" % cls.__name__,
            (AuditLog, Base),
            dict(
                __tablename__="%s_audit_log" % cls.__tablename__,
                parent_id=mapped_column(GUID, ForeignKey("%s.id" % cls.__tablename__)),
                audit_log=relationship(cls, viewonly=True),
            ),
        )
        # return relationship(cls.AuditLog, cascade="all, delete")
        # if not keeping audit log, remove log_delete event and user cascade delete
        # TODO: if not set cascade, enable background housekeeping for user choices
        return relationship(cls.AuditLog)

    @classmethod
    def log_create(cls, mapper, connection, target) -> None:
        target.audit_log.append(
            cls.AuditLog(request_id=request_id_ctx.get(), action="create", change_data=target.dict(), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get())
        )

    @classmethod
    def log_update(cls, mapper, connection, target) -> None:
        change_data = get_object_changes(target)
        target.audit_log.append(
            cls.AuditLog(request_id=request_id_ctx.get(), action="update", change_data=change_data), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get()
        )

    @classmethod
    def log_delete(cls, mapper, connection, target) -> None:
        target.audit_log.append(
            cls.AuditLog(request_id=request_id_ctx.get(), action="delete", change_data=target.dict()), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get()
        )

    @classmethod
    def __declare_last__(cls):
        # https://docs.sqlalchemy.org/en/14/orm/events.html
        # propagate=False¶ – When True, the event listener should be applied to all inheriting
        # mappers and/or the mappers of inheriting classes, as well as any mapper which is the target of this listener.
        event.listen(cls, "after_insert", cls.log_create, propagate=True)
        event.listen(cls, "after_update", cls.log_update, propagate=True)
        event.listen(cls, "after_delete", cls.log_delete, propagate=True)
```

**Comment by wangxin688** [comment]
by the way? which event.listen should I use for `update(obj)` in session level, seems it's not right when using `after_update` to catch update event. I know I can use 'after_execute' on engine level. Thanks a lot

**Reply by wangxin688** [reply]
> by the way? which event.listen should I use for `update(obj)` in session level, seems it's not right when using `after_update` to catch update event. I know I can use 'after_execute' on engine level. Thanks a lot

I have figured it out, replace with updating the object field one by one after query it rather than using update object directly.

**Reply by wangxin688** [reply]
> I want to record the db object change, so I created a Audit log mixin to record the database change to object to reduce a lot of duplicated code. I just check the examples you guys wrote before and made this mixin, but when I use it, there will be a warning generated. `SAWarning: Usage of the 'collection append' operation is not currently supported within the execution stage of the flush process. Results may not be consistent. Consider using alternative event listeners or connection-level operations instead. target.audit_log.append(` but the audit log was successfully add to database. should I just ignoring this message? or there is a better to way to do this? here's my code, I did not paste all of code here. (by the way, this code is also working under async session, amazing :--)
> 
> ```python
> import uuid
> from datetime import datetime
> from typing import TYPE_CHECKING, Dict, List, Optional
> 
> from sqlalchemy import DateTime, ForeignKey, String, event, inspect
> from sqlalchemy.dialects.postgresql import ARRAY, HSTORE, JSON
> from sqlalchemy.ext.declarative import declared_attr
> from sqlalchemy.orm import Mapped, class_mapper, mapped_column, relationship
> from sqlalchemy.orm.attributes import get_history
> from sqlalchemy.sql import func
> 
> from src.db.db_base import Base
> from src.db.db_types import GUID, GUID_GEN
> from src.register.context import auth_user_ctx, request_id_ctx, tenant_ctx
> 
> if TYPE_CHECKING:
>     from src.app.auth.models import User
> 
> 
> def get_object_changes(obj):
>     """Given a model instance, returns dict of pending
>     changes waiting for database flush/commit.
> 
>     e.g. {
>         'some_field': {
>             'before': *SOME-VALUE*,
>             'after': *SOME-VALUE*
>         },
>         ...
>     }
>     """
>     inspection = inspect(obj)
>     changes = {}
>     for attr in class_mapper(obj.__class__).column_attrs:
>         if getattr(inspection.attrs, attr.key).history.has_changes():
>             if get_history(obj, attr.key)[2]:
>                 before = get_history(obj, attr.key)[2].pop()
>                 after = getattr(obj, attr.key)
>                 if before != after:
>                     if before or after:
>                         changes[attr.key] = {"before": before, "after": after}
>     return changes
> 
> 
> class AuditLog:
>     id: Mapped[uuid.UUID] = mapped_column(primary_key=True, server_default=GUID_GEN)
>     created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=func.now(), nullable=False)
>     request_id: Mapped[str] = mapped_column(String(50), nullable=False, comment="request_context_id")
>     action: Mapped[str] = mapped_column(String(10), nullable=False, comment="literal action")
>     change_data: Mapped[dict] = mapped_column(JSON, nullable=False, comment="object change data")
> 
>     @declared_attr
>     def user_id(cls):
>         return mapped_column(
>             GUID, ForeignKey("auth_user.id", ondelete="SET NULL"), default=auth_user_ctx.get, nullable=True
>         )
> 
>     @declared_attr
>     def auth_user(cls) -> Mapped["User"]:
>         return relationship("User")
> 
> 
> class AuditLogMixin:
>     """Use generic association which persists association objects within
>     individual tables. each one generates to persist those objects on behalf of
>     a particular parent class.
>     """
> 
>     @declared_attr
>     def audit_log(cls) -> Mapped[List["AuditLog"]]:
>         cls.AuditLog = type(
>             "%sAuditLog" % cls.__name__,
>             (AuditLog, Base),
>             dict(
>                 __tablename__="%s_audit_log" % cls.__tablename__,
>                 parent_id=mapped_column(GUID, ForeignKey("%s.id" % cls.__tablename__)),
>                 audit_log=relationship(cls, viewonly=True),
>             ),
>         )
>         # return relationship(cls.AuditLog, cascade="all, delete")
>         # if not keeping audit log, remove log_delete event and user cascade delete
>         # TODO: if not set cascade, enable background housekeeping for user choices
>         return relationship(cls.AuditLog)
> 
>     @classmethod
>     def log_create(cls, mapper, connection, target) -> None:
>         target.audit_log.append(
>             cls.AuditLog(request_id=request_id_ctx.get(), action="create", change_data=target.dict(), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get())
>         )
> 
>     @classmethod
>     def log_update(cls, mapper, connection, target) -> None:
>         change_data = get_object_changes(target)
>         target.audit_log.append(
>             cls.AuditLog(request_id=request_id_ctx.get(), action="update", change_data=change_data), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get()
>         )
> 
>     @classmethod
>     def log_delete(cls, mapper, connection, target) -> None:
>         target.audit_log.append(
>             cls.AuditLog(request_id=request_id_ctx.get(), action="delete", change_data=target.dict()), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get()
>         )
> 
>     @classmethod
>     def __declare_last__(cls):
>         # https://docs.sqlalchemy.org/en/14/orm/events.html
>         # propagate=False¶ – When True, the event listener should be applied to all inheriting
>         # mappers and/or the mappers of inheriting classes, as well as any mapper which is the target of this listener.
>         event.listen(cls, "after_insert", cls.log_create, propagate=True)
>         event.listen(cls, "after_update", cls.log_update, propagate=True)
>         event.listen(cls, "after_delete", cls.log_delete, propagate=True)
> ```

after some investigation, i add a append method to auditlog,this will performs the flush operation explicitly. but it's still receive the same warning
```python3
from sqlalchemy.exc import InvalidRequestError

class AuditLog:
    # ...

    def append(self, obj) -> None:
        self.change_data.update(obj.change_data)
        try:
            session = inspect(self).session
            if session:
                session.add(obj)
                session.flush([obj])
        except InvalidRequestError:
            pass
```

**Comment by zzzeek** [comment, marked as answer]
if you want to manipulate the structure of related objects at flush time, you need to use flush-level events such as before_flush or [after_flush](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_flush).    if you want to emit log lines within mapper-level events like `after_update`, I would recommend use the connection that's right there to insert the log line immediately:

```py
   @classmethod
    def log_create(cls, mapper, connection, target) -> None:
         connection.execute(
              insert(cls.AuditLog),
              dict(user_id=cls.id, request_id=request_id_ctx.get(), action="create", change_data=target.dict(), parent_id=target.id, user_id=auth_user_ctx.get(), tenant_id=tenant_ctx.get())
         )
```

**Reply by wangxin688** [reply]
Thanks a lot. After trying this solution, problem solved.
